### DISCUSSION TITLE
v0.9.8

### DISCUSSION THREAD

**Discussion by Mimetis** [Mimetis/Dotmim.Sync#951] [General]
New version `v0.9.8` is available on [nuget](https://www.nuget.org/packages/Dotmim.Sync.Core/).

- New provider for Postgres using [Npgsql](https://www.npgsql.org/) ([Github repository](https://github.com/npgsql/npgsql)) : [Dotmim.Sync.PostgreSql](Dotmim.Sync.PostgreSql) : See https://github.com/Mimetis/Dotmim.Sync/discussions/951#discussioncomment-4819623
- Transient errors handling during applying phases: See https://github.com/Mimetis/Dotmim.Sync/discussions/951#discussioncomment-4819874
- Automatic Upgrade on both side (no need to call `remoteOrchestrator.NeedsToUpgradeAsync()` anymore)
- Performances improvements
- Handling errors and retry on error: See https://github.com/Mimetis/Dotmim.Sync/discussions/951#discussioncomment-4819913

**Comment by Mimetis** [comment]
# New Postgres provider

A new provider to sync **PostgreSQL** databases is now available: https://www.nuget.org/packages/Dotmim.Sync.PostgreSql/ 
Thanks to @WasimAhmad for initiating the work on that !

The `NpgsqlSyncProvider` provider is using the [Npgsql](https://www.npgsql.org/) **ADO.NET** provider behind the scene to access your **PostgreSQL** database.
`NpgsqlSyncProvider` is working only from **PostgreSQL** version > **15**

If you want to test PostgreSQL, you can use this demo [AdventureWorks script](https://github.com/Mimetis/Dotmim.Sync/blob/master/CreatePostgresAdventureWorks.sql) 

Then you can use this sample to sync your PostgreSQL database:

``` cs
using Dotmim.Sync.PostgreSql;

var serverProvider = new NpgsqlSyncProvider(serverConnectionString);
var clientProvider = new NpgsqlSyncProvider(clientConnectionString);
// Of course you can use a SQLite / SQL Server / MariaDB / MySQL client as well !
// var clientProvider = new SqliteSyncProvider("test.db");

var setup = new SyncSetup("ProductDescription", "ProductCategory",
 "ProductModel", "Product", "Address", "Customer", "CustomerAddress",
 "SalesOrderHeader", "SalesOrderDetail");

var options = new SyncOptions { DisableConstraintsOnApplyChanges=true };

var progress = new SynchronousProgress<ProgressArgs>(s =>
    Console.WriteLine($"{s.ProgressPercentage:p}:  \t[{s?.Source[..Math.Min(4, s.Source.Length)]}] {s.TypeName}: {s.Message}"));

var s = await agent.SynchronizeAsync(setup, progress: progress);
Console.WriteLine(s);
```
![image](https://user-images.githubusercontent.com/4592555/215506327-2d63ed40-fe62-4419-84fa-bc179fcc70c6.png)

![image](https://user-images.githubusercontent.com/4592555/215506792-44289a8e-73db-47f2-bad9-06993e10252a.png)

**Be careful**, as for now, the `NpgsqlSyncProvider` is not handling non common types, as array and so on... 
Please use it carefully and validate that all your columns types and tables are handled correctly by **DMS**

**Comment by Mimetis** [comment]
## Transient errors

Add new interceptor `OnRowsChangesFallbackFromBatchToSingleRowApplying` when an error in batch mode fallback to line per line. 
(Only available for providers supporting batch mode ... well only **SQL Server**!).

``` cs
localOrchestrator.OnRowsChangesFallbackFromBatchToSingleRowApplying(async args =>
{
 Console.WriteLine($"- Applying one by one because of an error occurred during batch applying");
 Console.WriteLine($"- Error raised causing the fallback: {args.Exception.Message}");
 Console.WriteLine($"- Rows count to apply one by one: {args.SyncRows.Count}");
});
```

Add new interceptor `OnTransientErrorOccured` when a transient error happens during applying changes on local data source.
Transient errors are usually **Deadlocks** or **Timeout**. 
``` cs
localOrchestrator.OnTransientErrorOccured(async args =>
{
 Console.WriteLine($"- Transient error happened: {this.HandledException?.Message}. Retry: {this.Retry}");
});
```
Add transients error codes for all providers.

**Comment by Mimetis** [comment]
## Errors handling

See https://dotmimsync.readthedocs.io/Errors.html

**Comment by carlosxjose** [comment]
Great job!!!

**Comment by VagueGit** [comment]
Tested on database syncing ~1million rows across 41 tables. Great work!
Not having to call `remoteOrchestrator.NeedsToUpgradeAsync()` is a big one for us.
`OnRowsChangesFallbackFromBatchToSingleRowApplying` is big for us.

**Reply by Mimetis** [reply]
for what reason `OnRowsChangesFallbackFromBatchToSingleRowApplying` is important ?

**Comment by VagueGit** [comment]
Monitoring sync across 600+ companies globally is a challenge. Any further insight into sync error context is welcome.

As I understand FallbackFromBatchToSingleRowApplying occurs when there is a sync error in a batch. Being able to monitor single rows applying gives context to the error.

**Comment by Rustam-23** [comment]
Thank you for this great work, @Mimetis!
Json column support by NpgsqlSyncProvider possible in the future?

**Reply by Mimetis** [reply]
I don't think so, as we don't have any corresponding column in SQLite, MySql, MariaDB, SQL Server
